src/port/port.tcl

```diff
@@ -3927,6 +3927,161 @@ proc action_echo { action portlist opts } {
+        set temp [file dirname $temp]
+    }
+
+    if {[llength $paths] > 0} {
+        set allow [$macports::ui_options(questions_yesno) "The 'macports' user needs search access to $path to use ports in
+this directory. The following directories will be changed to include 'x'
+permissions for group and others:\n" "allow" $paths {y} 0 "Do you want to allow this?"]
+        if {$allow == 1} {
+            # quit as user answered 'no'
+            set tempfd [open "${macports::portdbpath}/contrib" w]
+            close $tempfd
+            return 0
+        }
+        foreach path $paths {
+            file attributes $path -permissions [expr {[file attributes $path -permissions] | 1}]
+        }
+    }
+
+    # CREATE GIT SOURCE BY CLONING MACPORTS-PORTS
+    # The small depth is for performance (it also implies --single-branch)
+    set url https://github.com/macports/macports-ports.git
+    if {[catch {system -nodup "[macports::findBinary git] clone --depth 1 $url $path"} result]} {
+        ui_debug $::errorInfo
+        ui_error "git clone: $result"
+        return 1
+    }
+
+    # COPY INDEXFILE FROM OLD SOURCE TO GIT SOURCE
+    set default_source_url [lindex $sources_default 0]
+    set indexfile [macports::getindex $default_source_url]
+
+    if {[catch {file copy -force $indexfile $path} error]} {
+        ui_debug $::errorInfo
+        ui_error "file copy: $error"
+        return -code error "Copying PortIndex to new git-based source failed"
+    }
+
+    # EDIT SOURCES.CONF TO USE GIT SOURCE
+    if {![file exists $temp]} {
+        return -code error "sources.conf not found in $confpath"
+    }
+
+    # Construct sed command
+    set cmdline "[macports::findBinary sed] -E s/(^rsync)/#\1/g"
+    if {[catch {set tmpfile [mkstemp "/tmp/sources.conf.sed.XXXXXX"]} error]} {
+        ui_debug $::errorInfo
+        ui_error "mkstemp: $error"
+        return -code error "mkstemp failed"
+    }
+    set tmpfd [lindex $tmpfile 0]
+    set tmpfile [join [lrange $tmpfile 1 end]]
+    set attributes [file attributes $confpath]
+    lappend cmdline "<${confpath}" ">@$tmpfd"
+    # Comment out the rsync line
+    if {[catch {exec -ignorestderr -- {*}$cmdline} error]} {
+        ui_debug $::errorInfo
+        ui_error "sed: $error"
+        file delete "$tmpfile"
+        close $tmpfd
+        return -code error "sed sed(1) failed"
+    }
+
+    # Write the new source
+    puts $tmpfd "file://[file normalize $path] \[default\]"
+
+    if {[geteuid] != [getuid]} {
+        seteuid $oldeuid
+        setegid $oldgid
+    }
+
+    # Replace the old sources.conf with the new one
+    if {[catch {file rename -force $tmpfile $confpath} error]} {
+        ui_debug $::errorInfo
+        ui_error "move: $error"
+        file delete "$tmpfile"
+        return -code error "sources.conf overwrite failed"
+    }
+
+    # Restore sources.conf attributes
+    file attributes $confpath {*}$attributes
+
+    # Update the ports tree
+    try {
+        mportsync
```

Comment:
Need help with this one. At this point we've setup the new git source including updating sources.conf, but calling `mportsync` still updates the old source.

Reply:
I see two options:
 - Call `mportshutdown` and `mportinit` again, then run `mportsync`
 - Use `execl` to re-execute the same command after the changes were made, see https://github.com/macports/macports-base/blob/gsoc17-migrate/src/port/port.tcl#L2829